webui/react/src/components/ExperimentActions.tsx

```diff
@@ -0,0 +1,169 @@
+  const unarchiveButton = <Button key="unarchive" loading={buttonStates.archive}
+    type="primary" onClick={archiveCB(false)}>
+    Unarchive</Button>;
+
+  const forkButton = <Button disabled key="fork" type="primary">Fork</Button>;
+  const pauseButton = <Button key="pause" loading={buttonStates[RunState.Paused]}
+    type="primary" onClick={requestExpStateCB(RunState.Paused)}>
+    Pause</Button>;
+  const activateButton = <Button key="activate" loading={buttonStates[RunState.Active]}
+    type="primary" onClick={requestExpStateCB(RunState.Active)}>
+    Activate</Button>;
+
+  const cancelButton = <Popconfirm
+    cancelText="No"
+    key="cancel"
+    okText="Yes"
+    title="Are you sure you want to kill the experiment?"
+    onConfirm={requestExpStateCB(RunState.StoppingCanceled)}
+  >
+    <Button danger loading={buttonStates[RunState.StoppingCanceled]}
+      type="primary">
+    Cancel</Button>
+  </Popconfirm>;
+
+  const killButton = <Popconfirm
+    cancelText="No"
+    key="kill"
+    okText="Yes"
+    title="Are you sure you want to kill the experiment?"
+    onConfirm={killExperimentCB}
+  >
+    <Button danger loading={buttonStates.kill} type="primary">Kill</Button>
+  </Popconfirm>;
+
+  const tsbButton = <Button key="tensorboard"
+    loading={buttonStates.tsb} type="primary" onClick={launchTensorboardCB}>
+      Tensorboard</Button>;
+
+  interface ConditionalButton {
+    btn: React.ReactNode;
+    showIf?: (exp: ExperimentDetails) => boolean;
+  }
+
+  const experimentWillNeverHaveData = (experiment: ExperimentDetails): boolean => {
+    const isTerminal = terminalRunStates.has(experiment.state);
+    // with lack of step state we can use numSteps as a proxy to trials that definietly have some
+    // metric.
+    const trialsWithSomeMetric = experiment.trials.filter(trial => trial.numSteps > 1);
+    return isTerminal && trialsWithSomeMetric.length === 0;
+  };
+
+  const actionButtons: ConditionalButton[] = [
+    { btn: forkButton },
+    {
+      btn: archiveButton,
+      showIf: (exp): boolean => terminalRunStates.has(exp.state) && !exp.archived,
+    },
+    {
+      btn: unarchiveButton,
+      showIf: (exp): boolean => terminalRunStates.has(exp.state) && exp.archived,
+    },
+    {
+      btn: pauseButton,
+      showIf: (exp): boolean => exp.state === RunState.Active,
+    },
+    {
+      btn: activateButton,
+      showIf: (exp): boolean => exp.state === RunState.Paused,
+    },
+    {
+      btn: cancelButton,
+      showIf: (exp): boolean => cancellableRunStates.includes(exp.state),
+    },
+    {
+      btn: killButton,
+      showIf: (exp): boolean => killableRunStates.includes(exp.state),
+    },
+    {
+      btn: tsbButton,
+      showIf: (exp): boolean => !experimentWillNeverHaveData(exp),
+    },
+  ];
+
+  return (
+    <ul className={css.base}>
+      {actionButtons
+        .filter(ab => !ab.showIf || ab.showIf(experiment as ExperimentDetails))
+        .map(ab => ab.btn)
+      }
```

Comment:
**suggestion:** We talk more about this one async if you want. In short, this might have too many layers of abstraction. IMO just having all the JSX buttons listed out here with all the properties set will keep this file shorter and easier to read. For an example, check out this [PR](https://github.com/determined-ai/determined/pull/866/files), specifically `ExperimentList.tsx` lines 224-245. With the current setup, if you need to edit the `Cancel` button click handler you have to find `actionButtons` then look for `cancelButton` then finally `requestExpStateCB`.

Reply:
hmm let's chat about this.

Reply:
We discussed this async and decided the array filter and map still makes sense but decided to move the buttons inline into the `actionButtons` list to reduce vertical file space.